Fixes to xfem 3d cutter mesh (#32780)
idaholab/moose: modules/xfem/doc/content/source/userobjects/CrackMeshCut3DUserObject.md

- **Curved free surfaces.** Because the curved surface is tiled by many small flat polygons, the candidate naturally snaps to the polygon of the tiling that is closest to it; the snap point varies smoothly across the curvature.
- **Concave corners.** When the candidate is just past a corner edge shared between two adjacent faces, the edge-fallback gives the same point on that shared edge for both faces. The outward nudge then pushes the inactive node into the empty space outside the corner rather than burying it back inside either face.

### Minimum-displacement reuse

After the projection step, the inactive endpoint's actual displacement from its previous position to the new position is compared against half of the active neighbor's growth length. If

```
||new_position - previous_inactive_position|| < 0.5 * active_growth_length

> ```

Reviewer: This is an equation, not code or input syntax. MooseDocs renders math through katex, so write it as `$...$` inline or in a bare `\begin{equation}...\end{equation}` block rather than a plain fence; the fence also loses the norm bars and multiplication signs to a monospace font. Same for the relation in the fence at lines 30-32.

Author: Done. Both relations now use KaTeX equation blocks.

_Prepared with Codex, powered by GPT-5._

the inactive node is left at its current location and the existing node is reused for the new crack-front element. This avoids producing degenerate sliver triangles when projection clips the inactive motion almost back to where it started (for example, when an inactive endpoint is pressed against a wall and growth would only push it deeper). When this rule fires, the connection between the old front and the new front shares a node at that endpoint, and the zero-area sliver triangle that would otherwise be generated is dropped by the existing minimum-area filter.

!alert note

> The closest-point projection only considers surfaces tagged into a sideset. A free surface that exists geometrically but is not assigned to any sideset will be invisible to this projection, and an inactive endpoint pushed into the body through such an untagged surface will not be corrected. When meshing for crack-growth simulations, every exterior face that the cutter can interact with should be assigned to a sideset.

Reviewer: This alert does not match the implementation. `projectInteriorInactiveEndpoint` selects candidate faces with `if (elem->neighbor_ptr(side)) continue;` (modules/xfem/src/userobjects/CrackMeshCut3DUserObject.C:996), a purely topological exterior-face test, and the file makes no `BoundaryInfo` or boundary-ID query anywhere. Every geometric exterior face is considered whether or not it is tagged into a sideset, so this alert warns of a failure mode that cannot occur and asks the user to do meshing work that has no effect. Drop the alert, or replace it with the limitation that does exist: the search is restricted to the terminal element of the ray traversal and that element's point neighbors, and the candidate is returned unchanged when no exterior face is found in that patch.

Author: Updated the implementation to require a BoundaryInfo registration when selecting exterior faces. If no registered face exists in the local patch, or the outward nudge remains inside, the previous known-outside endpoint is retained. The alert now describes that behavior.

_Prepared with Codex, powered by GPT-5._

This class: (1) reads in a mesh describing the crack surface, (2) uses the mesh to do initial cutting of 3D elements, and (3) grows the mesh incrementally based on prescribed growth functions. The code is interfaced with domain integral methods to allow nonplanar crack growth based on empirical propagation direction and speed laws.

> For crack-growth problems, [!param](/UserObjects/CrackMeshCut3DUserObject/size_control) sets the target spacing used when advancing and refining the cutter front, while [!param](/UserObjects/CrackMeshCut3DUserObject/min_elem_area) prevents the growth algorithm from adding very small or nearly degenerate triangles to the cutter mesh. The number of growth advances performed in one update is controlled by [!param](/UserObjects/CrackMeshCut3DUserObject/n_step_growth).

Reviewer: The constraint this change now enforces is missing: when [!param](/UserObjects/CrackMeshCut3DUserObject/n_step_growth) is nonzero, `size_control` must be set by the user and positive, otherwise the constructor errors with "Crack growth needs a positive size_control." The prose gives the meaning but not the requirement.

```suggestion
For crack-growth problems, [!param](/UserObjects/CrackMeshCut3DUserObject/size_control) sets the target spacing used when advancing and refining the cutter front, while [!param](/UserObjects/CrackMeshCut3DUserObject/min_elem_area) prevents the growth algorithm from adding very small or nearly degenerate triangles to the cutter mesh. The number of growth advances performed in one update is controlled by [!param](/UserObjects/CrackMeshCut3DUserObject/n_step_growth); when it is nonzero, [!param](/UserObjects/CrackMeshCut3DUserObject/size_control) must be set to a positive value.
```

Author: Done.

_Prepared with Claude Opus 4.8 via Claude Code._

- **Curved free surfaces.** Because the curved surface is tiled by many small flat polygons, the candidate naturally snaps to the polygon of the tiling that is closest to it; the snap point varies smoothly across the curvature.
- **Concave corners.** When the candidate is just past a corner edge shared between two adjacent faces, the edge-fallback gives the same point on that shared edge for both faces. The outward nudge then pushes the inactive node into the empty space outside the corner rather than burying it back inside either face.

### Minimum-displacement reuse

After the projection step, the inactive endpoint's actual displacement from its previous position to the new position is compared against half of the active neighbor's growth length. If

\begin{equation}
\left\lVert \boldsymbol{x}_{\mathrm{new}} - \boldsymbol{x}_{\mathrm{previous}} \right\rVert

> < 0.5 \Delta a_{\mathrm{active}}.

Reviewer: The full stop ends the sentence in the middle of it: the clause opened by "If" above the equation continues after it with a lowercase "the inactive node is left at its current location". Close the equation with a comma.

```suggestion
< 0.5 \Delta a_{\mathrm{active}},
```

Author: Done.

_Prepared with Claude Opus 4.8 via Claude Code._

## Example Input Syntax

This example shows the `Mesh` block in [list:mesh] needed for creating the cutter mesh along with the `CrackMeshCut3DUserObject` block in [list:cutobject].  The mesh block in [list:mesh] defines two separate meshes.  The cutter mesh is created in the `read_in_cutter_mesh` block and must have [!param](/Mesh/FileMeshGenerator/save_with_name) set in order to specify this mesh in `CrackMeshCut3DUserObject` shown in [list:cutobject] using [!param](/UserObjects/CrackMeshCut3DUserObject/mesh_generator_name).  The mesh used by the FEM simulation is specifed in the `FEM_mesh` block in this example and [!param](/Mesh/MeshGeneratorMesh/final_generator)`=FEM_mesh` must be set because only the `FEM_mesh` will be used for the finite-element solution and the mesh created by `read_in_cutter_mesh` will be ignored by the solution.

> The surface-growth regression inputs prescribe compatible stress and strain material properties without solving for equilibrium so that [DomainIntegralAction.md] can drive cutter growth directly.

Reviewer: "The surface-growth regression inputs" has no antecedent on this page: neither of the two inputs listed below is one of them, and none of them are named, so the reader cannot tell which inputs the sentence describes. Name the input.

```suggestion
The surface-growth tests built on `face_crack_3d_domain.i` prescribe compatible stress and strain material properties and set `solve = false`, so that [DomainIntegralAction.md] can drive cutter growth directly.
```

Author: Done. Named face_crack_3d_domain.i as the input the surface-growth tests are built on.

_Prepared with Claude Opus 4.8 via Claude Code._

This class: (1) reads in a mesh describing the crack surface, (2) uses the mesh to do initial cutting of 3D elements, and (3) grows the mesh incrementally based on prescribed growth functions. The code is interfaced with domain integral methods to allow nonplanar crack growth based on empirical propagation direction and speed laws.

> For crack-growth problems, [!param](/UserObjects/CrackMeshCut3DUserObject/size_control) sets the target spacing used when advancing and refining the cutter front, while[!param](/UserObjects/CrackMeshCut3DUserObject/min_elem_area) prevents the growth algorithm from adding very small or nearly degenerate triangles to the cutter mesh. The number of growth advances performed in one update is controlled by [!param](/UserObjects/CrackMeshCut3DUserObject/n_step_growth).

Reviewer: There is a missing space before the second `[!param]` shortcode. The text "...cutter front, while[!param](...)..." renders with "while" run directly into the parameter name and link with no separating space.
```suggestion
For crack-growth problems, [!param](/UserObjects/CrackMeshCut3DUserObject/size_control) sets the target spacing used when advancing and refining the cutter front, while [!param](/UserObjects/CrackMeshCut3DUserObject/min_elem_area) prevents the growth algorithm from adding very small or nearly degenerate triangles to the cutter mesh. The number of growth advances performed in one update is controlled by [!param](/UserObjects/CrackMeshCut3DUserObject/n_step_growth).
```

Author: Done.

_Prepared with Codex, powered by GPT-5._

> ## Surface-Cutting Segment, Active and Inactive Nodes

Reviewer: The heading introduces "Surface-Cutting Segment", but the section never uses or defines that term -- the body calls them "active boundary segments". Name the concept the section actually describes.

```suggestion
## Active and Inactive Cutter-Boundary Nodes
```

Author: Done.

_Prepared with Codex, powered by GPT-5._

```
inactive_start — active_1 — active_2 — ... — active_N — inactive_end
```

Classification is done on every growth step by testing whether each cutter-boundary node lies inside the FEM body:

- A boundary node is **active** if it lies inside the body.
- A boundary node is **inactive** if it lies outside the body.

The crack front used by [DomainIntegralAction.md] and the propagation laws is composed only of active nodes. The inactive endpoints act as anchors for the active segment and define where the crack front meets the free surface.

## Motion of the Inactive Node

Each active node advances in the direction and by the increment determined by the growth law (function-prescribed or hoop-stress-based). The inactive endpoints are not driven by their own growth law; instead they inherit growth direction and increment from the active neighbor they are connected to.  So the candidate position for an inactive endpoint is

```
candidate = previous_inactive_position + active_direction * active_growth_length
```

> ### Keeping the inactive node above the free surface

Reviewer: "above the free surface" is ambiguous -- there is no privileged vertical direction here, and everywhere else the section states the condition as lying *outside the body* (lines 22, 36, 38). Use "outside the body" in this heading and in the same phrase on line 39 so the invariant reads the same way throughout.

```suggestion
### Keeping the inactive node outside the body
```

Author: Done.

_Prepared with Codex, powered by GPT-5._
